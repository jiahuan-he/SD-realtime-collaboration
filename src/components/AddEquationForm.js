import React from 'react';

const button = {
    backgroundColor: "white",
    border: "1px solid",
    color: "black",
    padding: "5px 12px",
    cursor: "pointer",
}

const inputInvalid = {
    border: "2px solid #ed6663"
}

export default class AddEquationForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: "",
            equation: "",
        };
    }

    handleChangeId = (event) => {
        this.setState({ id: event.target.value});
    }

    handleChangeEquation = (event) => {
        this.setState({ equation: event.target.value});
    }

    isValidId = () => {
        if(!this.state.id) return false
        
        if(this.props.stocks.filter((stock => stock.id === this.state.id)).length === 0
        && this.props.flows.filter((flow => flow.id === this.state.id)).length === 0
        ) return false
        
        return true 
    }

    isValidEquation = () => {
        if(!this.state.equation.trim()) return false        
        return true 
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <label>
                    Name
                <input 
                    style={this.isValidId()?null:inputInvalid} 
                    type="text" 
                    value={this.state.id} onChange={this.handleChangeId} />
                </label>
                <label>
                    Equation
                <input 
                    style={this.isValidEquation()?null:inputInvalid} 
                    type="text" 
                    value={this.state.equation} onChange={this.handleChangeEquation}/>
                </label>
                <input type="button" value={"Add Equation"} style={button}
                    onClick={() => {
                        if(!this.isValidId() || !this.isValidEquation()) return                        
                        this.props.addEquation(this.state.equation, this.state.id)
                    }}
                />
            </form>
        )
    }
}