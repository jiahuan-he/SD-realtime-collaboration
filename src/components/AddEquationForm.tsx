import React, { CSSProperties, FormEvent } from 'react';
import { I_Stock, I_Flow} from '../model';

interface State{
    id: string,
    equation:string,
}

interface Props{
    stocks: Array<I_Stock>,
    flows: Array<I_Flow>,
    addEquation: (equation:string, id:string)=>void,
    button: CSSProperties,
    inputInvalid: CSSProperties,
}


export default class AddEquationForm extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            id: "",
            equation: "",
        };
    }

    handleChangeId = (event: FormEvent<HTMLInputElement>) => {
        this.setState({ id: event.currentTarget.value});
    }

    handleChangeEquation = (event: FormEvent<HTMLInputElement>) => {
        this.setState({ equation: event.currentTarget.value});
    }

    isValidId = () => {
        if(!this.state.id) return false
        
        if(this.props.stocks.filter((stock => stock.id === this.state.id)).length === 0
        && this.props.flows.filter((flow => flow.id === this.state.id)).length === 0
        ) return false
        
        return true 
    }

    isValidEquation = () => {
        if(!this.state.equation.trim()) return false        
        return true 
    }

    render() {
        return (
            <form>
                <label>
                    Name
                <input 
                    style={this.isValidId()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.id} onChange={this.handleChangeId} />
                </label>
                <label>
                    Equation
                <input 
                    style={this.isValidEquation()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.equation} onChange={this.handleChangeEquation}/>
                </label>
                <input type="button" value={"Add Equation"} style={this.props.button}
                    onClick={() => {
                        if(!this.isValidId() || !this.isValidEquation()) return                        
                        this.props.addEquation(this.state.equation, this.state.id)
                        this.setState({
                            id: "",
                            equation: "",
                        })
                    }}
                />
            </form>
        )
    }
}