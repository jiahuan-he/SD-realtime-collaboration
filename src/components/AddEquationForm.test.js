
import React from 'react';
import { shallow } from 'enzyme';
import AddEquationForm from './AddEquationForm';
describe('AddEquationForm', () => {
    it('should render correctly in "debug" mode', () => {
        const props = {
            "stocks": [{
                "dependencies": ["tank1", "flow"],
                "equation": "-flow",
                "id": "tank1",
                "initValue": 100,
                "posX": 55,
                "posY": 217,
                "value": 100
            }, {
                "dependencies": ["tank2", "flow"],
                "equation": "+flow",
                "id": "tank2",
                "initValue": 0,
                "posX": 347,
                "posY": 241,
                "value": 0
            }, {
                "dependencies": ["tank3"],
                "equation": "",
                "id": "tank3",
                "initValue": 0,
                "posX": 11,
                "posY": 17,
                "value": 0
            },],
            "flows": [{
                "dependencies": ["rate"],
                "equation": "rate",
                "from": "tank1",
                "id": "flow",
                "to": "tank2"
            },],
            "arrows": [{
                "from": "rate",
                "to": "flow"
            }], "parameters": [{
                "name": "rate",
                "posX": 257,
                "posY": 351,
                "value": 1
            }],
        };
        const form = shallow(<AddEquationForm
            debug
        ></AddEquationForm>);
        expect(form).toMatchSnapshot();
    });
});
