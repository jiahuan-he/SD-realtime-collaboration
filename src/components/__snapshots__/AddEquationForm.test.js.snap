// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEquationForm should render correctly in "debug" mode 1`] = `
<form>
  <label>
    Name
    <input
      onChange={[Function]}
      style={
        Object {
          "border": "2px solid #ed6663",
        }
      }
      type="text"
      value=""
    />
  </label>
  <label>
    Equation
    <input
      onChange={[Function]}
      style={
        Object {
          "border": "2px solid #ed6663",
        }
      }
      type="text"
      value=""
    />
  </label>
  <input
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "border": "1px solid",
        "color": "black",
        "cursor": "pointer",
        "padding": "5px 12px",
      }
    }
    type="button"
    value="Add Equation"
  />
</form>
`;
